base/compiler/abstractinterpretation.jl

```diff
@@ -703,118 +707,161 @@ function argtype_tail(argtypes::Vector{Any}, i::Int)
+                    bty = typesubtract(bty, typeof(aty.val))
                 end
-                if isa(a, Slot)
-                    return Conditional(a, aty, aty)
+                return Conditional(b, aty, bty)
+            end
+            if isa(bty, Const) && isa(a, Slot)
+                if rt === Const(false)
+                    bty = Union{}
+                elseif rt === Const(true)
+                    aty = Union{}
+                elseif aty isa Type && isdefined(typeof(bty.val), :instance) # same for b
+                    aty = typesubtract(aty, typeof(bty.val))
                 end
-            elseif f === Core.Compiler.not_int
-                aty = argtypes[2]
-                if isa(aty, Conditional)
-                    ifty = aty.elsetype
-                    elty = aty.vtype
-                    if rt === Const(false)
-                        ifty = Union{}
-                    elseif rt === Const(true)
-                        elty = Union{}
-                    end
-                    return Conditional(aty.var, ifty, elty)
+                return Conditional(a, bty, aty)
+            end
+            if isa(b, Slot)
+                return Conditional(b, bty, bty)
+            end
+            if isa(a, Slot)
+                return Conditional(a, aty, aty)
+            end
+        elseif f === Core.Compiler.not_int
+            aty = argtypes[2]
+            if isa(aty, Conditional)
+                ifty = aty.elsetype
+                elty = aty.vtype
+                if rt === Const(false)
+                    ifty = Union{}
+                elseif rt === Const(true)
+                    elty = Union{}
                 end
+                return Conditional(aty.var, ifty, elty)
             end
         end
-        return isa(rt, TypeVar) ? rt.ub : rt
+    end
+    return isa(rt, TypeVar) ? rt.ub : rt
+end
+
+function abstract_call_unionall(argtypes::Vector{Any})
+    if length(argtypes) == 3
+        canconst = true
+        if isa(argtypes[3], Const)
+            body = argtypes[3].val
+        elseif isType(argtypes[3])
+            body = argtypes[3].parameters[1]
+            canconst = false
+        else
+            return Any
+        end
+        if !isa(body, Type) && !isa(body, TypeVar)
+            return Any
+        end
+        if has_free_typevars(body)
+            if isa(argtypes[2], Const)
+                tv = argtypes[2].val
+            elseif isa(argtypes[2], PartialTypeVar)
+                ptv = argtypes[2]
+                tv = ptv.tv
+                canconst = false
+            else
+                return Any
+            end
+            !isa(tv, TypeVar) && return Any
+            body = UnionAll(tv, body)
+        end
+        ret = canconst ? AbstractEvalConstant(body) : Type{body}
+        return ret
+    end
+    return Any
+end
+
+# call where the function is known exactly
+function abstract_call_known(interp::AbstractInterpreter, @nospecialize(f),
```

Comment:
Allocating all these Tuple objects seems unfortunate. We've tried to work pretty hard to remove and minimize that pattern, as it seems that it can sap performance.

Reply:
Should we do a Pair{Any, Any} or a specialized struct instead. Seems like that may potentially be easier on the compiler.

Reply:
Reminder to self: Whatever we do here, should also apply to e.g. typeinf_edge.

Reply:
@vtjnash does Pair{Any,Any} sound reasonable to you?